### PR TITLE
Refactor topology out of token_metadata

### PR DESCRIPTION
**Pull request by bhalevy**
This series moves the topology code from locator/token_metadata.{cc,hh} out to localtor/topology.{cc,hh}
and introduces a shared header file: locator/types.hh contains shared, low level definitions, in anticipation of https://github.com/scylladb/scylladb/pull/11987

While at it, the token_metadata functions are turned into coroutines
and topology copy constructor is deleted.  The copy functionality is moved into an async `clone_gently` function that allows yielding while copying the topology.

### CODE REVIEW

File: locator/token_metadata.cc [hunk 1]
```diff
@@ -1261,13 +1259,33 @@ topology::topology(config cfg)
...
 }
 
-topology::topology(const topology& other)
-        : _dc_endpoints(other._dc_endpoints)
-        , _dc_racks(other._dc_racks)
-        , _current_locations(other._current_locations)
-        , _pending_locations(other._pending_locations)
-        , _sort_by_proximity(other._sort_by_proximity)
-{
+future<topology> topology::clone_gently() const {
```
Comment (xemul) [inline locator/token_metadata.cc]: suggestion for incremental change -- worth extending the util/stall_free.hh with cloning?

File: locator/types.hh [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+/*
+ * SPDX-License-Identifier: (AGPL-3.0-or-later and Apache-2.0)
+ */
+
+#pragma once
+
+#include <seastar/core/sstring.hh>
+
+#include "inet_address_vectors.hh"
```
Comment (xemul) [inline locator/types.hh:15]: This one looks unrelated

File: locator/topology.hh [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+#include "locator/types.hh"
+
+using namespace seastar;
+
+namespace locator {
+
+class topology {
+public:
+    using inet_address = gms::inet_address;
```
Comment (xemul) [inline locator/topology.hh:28]: nit: this thing appeared during move, but there's namespace locator -wide declaration under the same name. Is this really required here?

**Comment by xemul** [comment]
LGTM modulo several nitpicks

File: locator/token_metadata.cc (hunk 1 above)
Comment (bhalevy) [reply locator/token_metadata.cc]: care to provide a more specific example? I don't follow exactly where you're aiming to...

File: locator/topology.hh (hunk 3 above)
Comment (bhalevy) [reply locator/topology.hh:28]: It previously relied on https://github.com/scylladb/scylladb/blob/d85f731478211c3952114247807eb424bb400b62/locator/token_metadata.hh#L37
I can move it too to locator/types.hh  but I don't like it very much, hence I restricted it to class topology.

Comment (bhalevy) [reply locator/topology.hh:28]: Funny thing, is that I actually did do that eventually and kept both.
So I'll remove this one.  Should be unneeded.

File: locator/types.hh (hunk 2 above)
Comment (bhalevy) [reply locator/types.hh:15]: hmm, I think I just copied it from token_metadata.hh, but including gms/inet_address.hh should be sufficient

Comment (bhalevy) [reply locator/types.hh:15]: it was needed for inet_address_vector_replica_set in topology.hh, but it should indeed be included in topology.hh, not in types.hh

**Comment by bhalevy** [comment]
In v2 ([d0bd305](https://github.com/scylladb/scylladb/commit/d0bd305d162dccc13c474404a68f58686e4425a7)):
- include only gms/inet_address.hh in types.hh
  - #include "inet_address_vectors.hh" in topology.hh instead
- remove `using inet_address = gms::inet_address` from class topology

**Comment by asias** [comment]
Looks good.

File: locator/token_metadata.cc (hunk 1 above)
Comment (xemul) [reply locator/token_metadata.cc]: There's a bunch of polymorphic helpers in this header that can gently clean pretty much anything iterable. My suggestion is that it might makes sense to have gentle cloning helpers in this file.

Comment (bhalevy) [reply locator/token_metadata.cc]: interesting idea, sounds feasible
